### PR TITLE
CPP: Improve our generated code detection.

### PR DESCRIPTION
**Pull request by geoffw0**
Improve our heuristic for detecting generated C/C++ source files.  Simply finding the phrase "generated by" was responsible for most of our results before but was fairly prone to false positives.  The new heuristic is more complicated but considerably more robust in a number of ways.

Diff query on 35 projects:
 - lost: https://lgtm.com/query/3886192986863929904/
 - added: https://lgtm.com/query/4434651052766736778/

Unfortunately the new version is at least 50% slower, due mainly to the `concat`.  If this turns out to be a problem we can go back to analysing each comment separately, with (I think) only a small loss of accuracy for the fairly rare cases resembling:
```
// Generated
// by some such program.
```

### CODE REVIEW

**Comment by geoffw0** [comment]
The language test failure is `semmlecode-cpp-tests/query-tests/filters/ClassifyFiles/ClassifyFiles.qlref`, an internal test.  There's a corresponding PR for it at https://git.semmle.com/Semmle/code/pull/30122.

File: cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll [hunk 1]
```diff
@@ -3,11 +3,45 @@ import semmle.code.cpp.File
...
+      // `auto-generated`.
+      cond.replaceAll("generated", "(auto[\\w-]*[\\s/\\*\\r\\n]*)?generated")
+
+      // replace `!` with a regexp for end-of-sentence / separator characters.
+      .replaceAll("!", "[\\.\\?\\!\\-\\;\\,]")
+
+      // replace ` ` with a regexp for one or more whitespace characters
+      // (including newlines and `/*`).
+      .replaceAll(" ", "[\\s/\\*\\r\\n]+") +
+      ").*"
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll]: As a long-time Perl programmer, I am in awe of this use of a regex to transform another regex!

File: cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll [hunk 2]
```diff
@@ -36,12 +112,13 @@ predicate hasPragmaDifferentFile(File f) {
...
   cached AutogeneratedFile() {
-    exists(int limit, int head |
-      head <= 5 and
-      limit = max(int line | locations_default(_, underlyingElement(this), head, _, line, _)) + 5
-    |
-      exists (Comment c | c.getFile() = this and c.getLocation().getStartLine() <= limit and isAutogeneratedComment(c))
-    )
-    or hasPragmaDifferentFile(this)
+    autogeneratedComment(
+      concat(Comment c |
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll]: I got some speed-up out of changing `concat` to `strictconcat`. In general it's always faster to use the strict aggregates.

File: cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll]: It's actually only a string substitution on a regexp string, not a regexp on a regexp.  Things were getting surprisingly unreadable until I did it this way!

File: cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll]: Good idea, I'd be surprised if it makes a big difference in this case but it shouldn't do any harm!

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll]: It does, in fact, appear to make quite a big difference (done).

**Comment by jbj** [comment]
LGTM. I'll merge this together with the internal PR when the internal PR has been updated to point to the latest commit.

**Pull request merged**